test/bulk_schedule_test.cpp

```diff
@@ -48,3 +49,39 @@ TEST(bulk, bulk_transform) {
         EXPECT_EQ(i, output[i]);
     }
 }
+
+TEST(bulk, cancellation) {
+    unifex::single_thread_context ctx;
+    auto sched = ctx.get_scheduler();
+
+    const std::size_t count = 1000;
+
+    std::vector<int> output(count, 0);
+    // Cancel after two chunks
+    // For the serial implementation this will stop the third chunk onwards from
+    // being dispatched.
+    const std::size_t compare_index = unifex::bulk_cancellation_chunk_size*2 - 1;
+
+    // Bulk, but sequential to test strict cancellation of later work
+
+    unifex::sync_wait(
+        unifex::let_with_stop_source([&](unifex::inplace_stop_source& stopSource) {
+            return unifex::bulk_join(
+                unifex::bulk_transform(
+                    unifex::bulk_schedule(sched, count),
+                    [&](std::size_t index) noexcept {
+                        // Stop after second chunk
+                        if(index == compare_index) {
+                            stopSource.request_stop();
+                        }
+                        output[index] = index;
+                    }, unifex::seq));
+        }));
+
+    for (std::size_t i = 0; i <= compare_index; ++i) {
+        EXPECT_EQ(i, output[i]);
+    }
+    for (std::size_t i = compare_index+1; i < count; ++i) {
+        EXPECT_EQ(0, output[i]);
```

Comment:
I wonder whether we should provide this guarantee for `bulk_schedule()` that, at least for sequential execution policy, that the indices are delivered in ascending order.

Would a valid implementation of `bulk_schedule()` be allowed to deliver the indices in a different order which might break the assumptions of this test?

Reply:
There are questions around this that we should discuss.

Delivering the indices in some order is one question. Launching the tasks in some order is another - which may or may not be tied to the index. I don't think it matters for the purposes of this test, it is definitely a question we should bring up for standardisation.

I don't know if there is an answer with the current bulk algorithms either, but to be fair there are many unanswered questions there.